interfaces/builtin/shared_memory.go

```diff
@@ -28,36 +28,74 @@ import (
 
 	"github.com/snapcore/snapd/interfaces"
 	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/snap"
 )
 
 const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
 
-// The plug side of shared-memory implements auto-connect to a matching slot -
-// this is permitted even though the interface is super-privileged because using
-// a slot requires a store declaration anyways so just declaring a plug will not
-// grant access unless a slot was also granted at some point.
+// The plug side of shared-memory can operate in two modes: if the
+// private attribute is set to true, then it can be connected to the
+// implicit system slot to be given a private version of /dev/shm.
+//
+// For a plug without that attribute set, it will connect to a
+// matching application snap slot - this is permitted even though the
+// interface is super-privileged because using a slot requires a store
+// declaration anyways so just declaring a plug will not grant access
+// unless a slot was also granted at some point.
 const sharedMemoryBaseDeclarationPlugs = `
   shared-memory:
     allow-installation: true
     allow-connection:
-      slot-attributes:
-        shared-memory: $PLUG(shared-memory)
+      -
+        plug-attributes:
+          private: false
+        slot-attributes:
+          shared-memory: $PLUG(shared-memory)
+      -
+        plug-attributes:
+          private: true
+        slot-snap-type:
+          - core
```

Comment:
for this alternation, do we want to also enforce that shared-memory is empty? i.e. this currently says that the plug can be connected with the slot is from the core/snapd snap and when the plug has `private: true` but does not otherwise preclude connection via the base-declaration if shared-memory is set to something (though we check it later so perhaps it's okay to not have this in the base declaration)

Reply:
We have checks in `BeforePreparePlug` to ensure a plug with `private: true` set does not have a shared-memory attribute. So a plug like that will be discarded during the sanitize process before it could ever be a candidate for connection.

We could add something like:

    slot-attributes:
      shared-memory: $MISSING

... but slots from app snaps are never going to be candidates for this branch of the declaration, and the implicit slot is always added with no attributes.

Reply:
Sure the check in BeforePreparePlug is fine too, just sometimes it is also nice to have the base declaration be more specific to prevent foot shooting, but then again the base declaration can get complex to read so less is more there sometimes